File: algorithm/leetcodeObjectiveC/kth-smallest-element-in-a-sorted-matrix.m
// 378. Kth Smallest Element in a Sorted Matrix
// Difficulty: Medium

// Given a n x n matrix where each of the rows and columns are sorted in ascending order, find the kth smallest element in the matrix.

// Note that it is the kth smallest element in the sorted order, not the kth distinct element.

// Example:

// matrix = [
//    [ 1,  5,  9],
//    [10, 11, 13],
//    [12, 13, 15]
// ],
// k = 8,

// return 13.
// Note:
// You may assume k is always valid, 1 ≤ k ≤ n2.

// Time:  O(k * log(min(n, m, k))), with m x n matrix
// Space: O(min(n, m, k))

#import <Foundation/Foundation.h>

#pragma mark PriorityQueue

@interface PriorityQueue : NSObject

- (instancetype)initIsMinHeap:(BOOL)isMinHeap;
- (instancetype)init;
- (id)top;
- (id)pop;
- (void)push:(id)element;
-(NSInteger)count;
- (BOOL)isEmpty;

@end

#pragma mark Solution

int kthSmallest(NSArray* matrix, int k) {
  int kthSmallest = 0;
  PriorityQueue* minHeap = [[PriorityQueue alloc] init];
  [minHeap push:@[matrix[0][0], @(0), @(0)]];
   
  while (![minHeap isEmpty] && k--) {
    NSArray* top = [minHeap pop];
    kthSmallest = [top[0] intValue];
    [minHeap push:@[matrix[[top[1] intValue]][[top[2] intValue] + 1], top[1], @([top[2] intValue] + 1)]];
    if ([top[2] intValue] == 0) {
      [minHeap push:@[matrix[[top[1] intValue] + 1][0], @([top[1] intValue] + 1), @0]];
    }
  }
  return kthSmallest;
}

File: data-structure-and-algorithm/typescript/leetcode/src/kth-smallest-element-in-a-sorted-matrix.ts
/*
Given an n x n matrix where each of the rows and columns are sorted in ascending order, return the kth smallest element in the matrix.

Note that it is the kth smallest element in the sorted order, not the kth distinct element.



Example 1:

Input: matrix = [[1,5,9],[10,11,13],[12,13,15]], k = 8
Output: 13
Explanation: The elements in the matrix are [1,5,9,10,11,12,13,13,15], and the 8th smallest number is 13
Example 2:

Input: matrix = [[-5]], k = 1
Output: -5


Constraints:

n == matrix.length
n == matrix[i].length
1 <= n <= 300
-109 <= matrix[i][j] <= 109
All the rows and columns of matrix are guaranteed to be sorted in non-decreasing order.
1 <= k <= n2
*/

function kthSmallest(grid: number[][], kth: number): number {
  const [rows, cols] = [grid.length, grid[0].length];
  let left = grid[0][0];
  let right = grid[rows - 1][cols - 1] + 1;
  while (left < right) {
    const mid = (left + right) >> 1;
    let count = 0;
    for (let i = 0, j = cols - 1; i < rows; ++i) {
      for (; j >= 0 && grid[i][j] > mid; --j);
      count += j + 1;
    }
    if (count < kth) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  return left;
}
